core/templates/pages/skill-editor-page/navbar/skill-editor-navbar.directive.ts

```diff
@@ -145,7 +145,9 @@ angular.module('oppia').directive('skillEditorNavbar', [
           ctrl.$onInit = function() {
             $scope.skill = SkillEditorStateService.getSkill();
             $scope.validationIssues = [];
-            $scope.$on(EVENT_SKILL_INITIALIZED, _validateSkill);
+            SkillEditorStateService.getEventSkillInitializedSubject().subscribe(
```

Comment:
While using subjects might make sense in general, I thought the point of this PR was to remove the need for the event altogether, by having the skill editor service store the skill info and the controller/directive retrieve a reference to it? The pattern you've provided here seems reasonable in the general case (though maybe add me as reviewer to other PRs which use it so that we can make sure we're going in the right direction), but for this specific case I think that using the Subject infrastructure (or events at all) might be more convoluted than necessary, given what the code is doing.

Reply:
@seanlip, Got it! I will make the changes.
I have a couple more questions:
Q1>If this were angular 8, I would be in favor of removing this `$scope.validationIssues = [];` and `$scope.skill`.
Instead, I would add an `@Input()` to skill-navbar-directive and questions-tab-directive, and keep the skill object in the controller itself. That way no event will be required. Any change to the object will be passed on.

Is there is an equivalent of this `@Input()` in angularjs? 

Q2> How do we plan to notify the directive of changes in the object? (Using watch could complicate the migration process).
> by having the skill editor service store the skill info and the controller/directive retrieve a reference to it?
My idea would be to use subjects but what do you suggest I use.

Reply:
Q1) I don't know how @Input() works, sorry. This is something you might need to ask @bansalnitish. 

Q2) Ideally you set it up so that you don't have to notify the directive. You can do this by having the service return a reference that is automatically updated when the value is changed. If I remember correctly, one way we used to do this was with a `var data = {currentSkill: ...}` dict and return data.currentSkill. You'll need to check that, if data.currentSkill is subsequently changed, the resulting value is too. (That said ... we might need to figure out how to update the derived data too -- maybe we can link it to the data dict as well and provide similar getters in the API.)

Can you explain why a watch would complicate the migration process?

Also, one thought I had while writing this is that an issue we have is that we don't know what this would, ideally, look like in Angular 8. So I guess I have a question for you, which is: if you were implementing this functionality from scratch, so that it behaves the same way as the code here (but not necessarily following the same coding patterns), how would you do it? If we know the desired end result then we can chart a pathway from where we are now to that end result.

Reply:
> Can you explain why a watch would complicate the migration process?

The $watch has no equivalent in Angular, we will have to look for workarounds/ deal with it case by case. It would be better not to introduce more of watch.
> Also, one thought I had while writing this is that an issue we have is that we don't know what this would, ideally, look like in Angular 8. So I guess I have a question for you, which is: if you were implementing this functionality from scratch, so that it behaves the same way as the code here (but not necessarily following the same coding patterns), how would you do it? If we know the desired end result then we can chart a pathway from where we are now to that end result.

In this particular case, we would be using what I asked about in the first question in my previous comment. @Input, i.e. property binding... There will be a stackblitz link, **in the end**, showcasing what I think would be ideal.
In the stackblitz app there are two (important) components, app.component, and the-nav-bar.component

In the-nav-bar.component.ts, I have put this
```
  @Input() skill = {
    name: 'Placeholder Skill Name',
    value: 'Placeholder Skill Value'
  };
```
So what is will basically do is that it will open a property in the-nav-bar selector so you can pass in data to that component from the parent component (the component where the tag <the-nav-bar> is used in HTML). 

In order to pass data, you will have to bind the property by using square brackets []. I have done that in app.component.html,
```
<the-nav-bar [skill] = "skillObj"></the-nav-bar>
```
In app.component.ts I have made a  skillObj:
```
skillObj = {
    name: 'Skill Name',
    value: 'Skill Value'
  };
```
and in the same file I have put two ways if changing the object:
One through setTimeout and the other through a button,
```
ngOnInit() {
    setTimeout(
      () => {
        this.skillObj.name = 'Skill name Changed';
      }, 4000
    );
  }
  changeSkillValue() {
    this.skillObj.value = 'skill value changed';
  }
```

So within 4 seconds of the app loading, it will change the skill name to `skill name changed` and if you click a button saying `Click here to change skill value` it will change the skill value to `skill value changed`.

**Please note that I haven't explicitly made any attempt to reach the-nav-bar.component to inform it of changes happening in the skillObj, it happens automatically because of angular change detection, which is highly optimized**.

StackBlitz link: https://stackblitz.com/edit/angular-g5e6pl

Reply:
+cc @bansalnitish ptal at my comment and let me know if you agree or if you would go for a different way.